src/kvstore/Part.cpp

```diff
@@ -254,24 +256,60 @@ bool Part::commitLogs(std::unique_ptr<LogIterator> iter) {
             break;
         }
         default: {
-            LOG(FATAL) << "Unknown operation: " << static_cast<uint8_t>(log[0]);
+            LOG(FATAL) << idStr_ << "Unknown operation: " << static_cast<uint8_t>(log[0]);
         }
         }
 
         ++(*iter);
     }
 
     if (lastId >= 0) {
-        std::string commitMsg;
-        commitMsg.reserve(sizeof(LogID) + sizeof(TermID));
-        commitMsg.append(reinterpret_cast<char*>(&lastId), sizeof(LogID));
-        commitMsg.append(reinterpret_cast<char*>(&lastTerm), sizeof(TermID));
-        batch->put(folly::stringPrintf("%s%d", kCommitKeyPrefix, partId_), commitMsg);
+        if (putCommitMsg(batch.get(), lastId, lastTerm) != ResultCode::SUCCEEDED) {
+            LOG(ERROR) << idStr_ << "Commit msg failed";
+            return false;
+        }
     }
-
     return engine_->commitBatchWrite(std::move(batch)) == ResultCode::SUCCEEDED;
 }
 
+std::pair<int64_t, int64_t> Part::commitSnapshot(const std::vector<std::string>& rows,
+                                                 LogID committedLogId,
+                                                 TermID committedLogTerm,
+                                                 bool finished) {
+    auto batch = engine_->startBatchWrite();
```

Comment:
when previouse snapshot is under installing, we should check if this is the same snapshot and resume it, otherwise drop previous snapshot.

As described in the paper:

> * If lastIndex is larger than lastest snapshot's, save snapshot file and Raft state(lastIndex, lastTerm, lastConfig). discard any existing or partial snapshot.
> * If existing log entry has same index and term as lastIndex and lastTerm,discard log up through lastIndex(but retain any following entries) and reply.
>

Reply:
Currently,  if the leader changed when sending snapshot,  the new leader's snapshot request will be rejected due to lastTotalCount and lastTotalSize not matched.  And after some while, the follower will clean up all data and new leader will send snapshot again.